File: src/vm/memory.rs
use miette::Result;

use crate::{
    parser::ast::{DataSize, Immediate},
    vm::Error,
};

#[derive(Debug)]
pub struct Memory {
    pub(crate) storage: Vec<u8>,
}

impl Memory {
    pub fn new(size: usize) -> Self {
        Self {
            storage: vec![0; size],
        }
    }

    pub fn len(&self) -> usize {
        self.storage.len()
    }

    pub fn read(&self, addr: usize, size: DataSize) -> Result<Immediate> {
        match size {
            DataSize::Byte => {
                if addr + 1 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::Byte(self.storage[addr]))
            }
            DataSize::Word => {
                if addr + 2 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::Word(u16::from_le_bytes(
                    self.storage[addr..addr + 2].try_into().unwrap(),
                )))
            }
            DataSize::DWord => {
                if addr + 4 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::DWord(u32::from_le_bytes(
                    self.storage[addr..addr + 4].try_into().unwrap(),
                )))
            }
            DataSize::QWord => {
                if addr + 8 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::QWord(u64::from_le_bytes(
                    self.storage[addr..addr + 8].try_into().unwrap(),
                )))
            }
            DataSize::Float => {
                if addr + 4 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::Float(f32::from_le_bytes(
                    self.storage[addr..addr + 4].try_into().unwrap(),
                )))
            }
            DataSize::Double => {
                if addr + 8 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                Ok(Immediate::Double(f64::from_le_bytes(
                    self.storage[addr..addr + 8].try_into().unwrap(),
                )))
            }
        }
    }

    pub fn write(&mut self, addr: usize, value: Immediate, size: DataSize) -> Result<()> {
        match size {
            DataSize::Byte => {
                if addr + 1 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr] = value.as_u8()?;
            }
            DataSize::Word => {
                if addr + 2 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr..addr + 2].copy_from_slice(&value.as_u16()?.to_le_bytes());
            }
            DataSize::DWord => {
                if addr + 4 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr..addr + 4].copy_from_slice(&value.as_u32()?.to_le_bytes());
            }
            DataSize::QWord => {
                if addr + 8 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr..addr + 8].copy_from_slice(&value.as_u64()?.to_le_bytes());
            }
            DataSize::Float => {
                if addr + 4 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr..addr + 4].copy_from_slice(&value.as_f32()?.to_le_bytes());
            }
            DataSize::Double => {
                if addr + 8 > self.storage.len() {
                    return Err(Error::InstructionPointerOutOfBounds(addr).into());
                }
                self.storage[addr..addr + 8].copy_from_slice(&value.as_f64()?.to_le_bytes());
            }
        }
        Ok(())
    }
}

File: src/vm/Memory.zig
const std = @import("std");
const mem = std.mem;
const Allocator = mem.Allocator;
const ArrayList = std.array_list.Managed;
const DataSize = @import("../parser/immediate.zig").DataSize;
const Immediate = @import("../parser/immediate.zig").Immediate;

const Memory = @This();

storage: ArrayList(u8),

pub fn init(size: usize, allocator: Allocator) !Memory {
    var storage = try ArrayList(u8).initCapacity(allocator, size);
    try storage.appendNTimes(0x00, size);
    return Memory{ .storage = storage };
}

pub fn deinit(self: *Memory) void {
    self.storage.deinit();
}

pub fn len(self: *Memory) usize {
    return self.storage.items.len;
}

pub fn read(self: *Memory, addr: usize, size: DataSize) !Immediate {
    switch (size) {
        .byte => {
            if (addr + 1 > self.len()) return error.AddressOutOfBounds;
            return .{ .byte = self.storage.items[addr] };
        },
        .word => {
            if (addr + 2 > self.len()) return error.AddressOutOfBounds;
            return .{
                .word = mem.readInt(u16, self.storage.items[addr .. addr + 2][0..2], .little),
            };
        },
        .dword => {
            if (addr + 4 > self.len()) return error.AddressOutOfBounds;
            return .{
                .dword = mem.readInt(u32, self.storage.items[addr .. addr + 4][0..4], .little),
            };
        },
        .qword => {
            if (addr + 8 > self.len()) return error.AddressOutOfBounds;
            return .{
                .qword = mem.readInt(u64, self.storage.items[addr .. addr + 8][0..8], .little),
            };
        },
        .float => {
            if (addr + 4 > self.len()) return error.AddressOutOfBounds;
            const bytes = self.storage.items[addr .. addr + 4][0..4];
            return .{
                .float = @bitCast(mem.readInt(u32, bytes, .little)),
            };
        },
        .double => {
            if (addr + 8 > self.len()) return error.AddressOutOfBounds;
            const bytes = self.storage.items[addr .. addr + 8][0..8];
            return .{
                .double = @bitCast(mem.readInt(u64, bytes, .little)),
            };
        },
    }
}

pub fn write(self: *Memory, addr: usize, value: Immediate, size: DataSize) !void {
    switch (size) {
        .byte => {
            if (addr + 1 > self.len()) return error.AddressOutOfBounds;
            self.storage.items[addr] = value.asU8();
        },
        .word => {
            if (addr + 2 > self.len()) return error.AddressOutOfBounds;
            @memcpy(self.storage.items[addr .. addr + 2], &mem.toBytes(value.asU16()));
        },
        .dword => {
            if (addr + 4 > self.len()) return error.AddressOutOfBounds;
            @memcpy(self.storage.items[addr .. addr + 4], &mem.toBytes(value.asU32()));
        },
        .qword => {
            if (addr + 8 > self.len()) return error.AddressOutOfBounds;
            @memcpy(self.storage.items[addr .. addr + 8], &mem.toBytes(value.asU64()));
        },
        .float => {
            if (addr + 4 > self.len()) return error.AddressOutOfBounds;
            @memcpy(self.storage.items[addr .. addr + 4], &mem.toBytes(value.asF32()));
        },
        .double => {
            if (addr + 8 > self.len()) return error.AddressOutOfBounds;
            @memcpy(self.storage.items[addr .. addr + 8], &mem.toBytes(value.asF64()));
        },
    }
}
